### PR TITLE
Check semi-invulnerability before immunities, abilities, etc.

### PR DESCRIPTION
**Pull request by MacChaeger**
This mostly matters for stuff like Magic Bounce and Volt Absorb, which should not activate when the target of the move that would activate them is semi-invulnerable. You can thank Metronome Battle for me finding this.

### CODE REVIEW

**Comment by Marty-D** [comment]
Looks good! Could you change the event name to `TryImmunity`? More consistent with the other Try stuff.

File: dev-tools/globals.ts [hunk 1]
```diff
@@ -177,16 +177,18 @@ interface EventMethods {
...
 	onAnyDragOut?: (this: Battle, pokemon: Pokemon) => void
 	onAnySetStatus?: (this: Battle, status: Status, pokemon: Pokemon) => void
 	onAttract?: (this: Battle, target: Pokemon, source: Pokemon, effect: UnknownEffect) => void
 	onAccuracy?: (this: Battle, accuracy: number, target: Pokemon, source: Pokemon, move: Move) => number | boolean | null | void
 	onBasePower?: (this: Battle, basePower: number, pokemon: Pokemon, target: Pokemon, move: Move) => void
+	onBeforeImmunity?: (this: Battle, target: Pokemon, source: Pokemon, move: Move) => void
 	onBeforeSwitchIn?: (this: Battle, pokemon: Pokemon) => void
-	onBeforeMove?: (this: Battle, attacker: Pokemon, defender: Pokemon, move: Move) => void
 	onBeforeSwitchOut?: (this: Battle, pokemon: Pokemon) => void
+	onBeforeMove?: (this: Battle, attacker: Pokemon, defender: Pokemon, move: Move) => void
```
Comment (Marty-D) [inline dev-tools/globals.ts]: You moved this in the wrong direction!

**Comment by MacChaeger** [comment]
uh...

**Comment by Zarel** [comment]
I re-ran the test, it passes now.

**Comment by Marty-D** [comment]
:+1:

**Comment by Zarel** [comment]
Blah, which perfect accuracies do and don't hit semi-invulnerable pokemon?

I'm pretty sure Aerial Ace et al don't. What about Thunder in rain?

**Pull request merged**